Allow resetting time in google_travel_time

eifinger opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->

Allow resetting already defined departure or arrival times.
Equivalent to https://github.com/home-assistant/core/pull/88253

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

eifinger committed:
Use vol.Maybe

balloob commented:
Marked as draft until frontend is done.

eifinger committed:
Use add_suggested_values_to_schema

eifinger committed:
Fix tests

eifinger commented:
@epenet you reviewed the PR for waze. Could you also have a look on this one since it is basically the same

epenet reviewed (commented):
I'm not sure your PR works as intended. I think you may still need a default=None in some places.

Basically, when a blank string is selected in the frontend, the data passed back will not contain the key at all, and therefore will not update the options.
I think you may need to set `default=None` or `default=""` where you want the key to be present in the resulting options.

Please adjust the test to ensure other options can also still be cleared.

eifinger commented:
> I'm not sure your PR works as intended. I think you may still need a default=None in some places.
> 
> Basically, when a blank string is selected in the frontend, the data passed back will not contain the key at all, and therefore will not update the options. I think you may need to set `default=None` or `default=""` where you want the key to be present in the resulting options.
> 
> Please adjust the test to ensure other options can also still be cleared.

The only thing which can be cleared is `CONF_TIME` and that works. I checked that manually and the tests do reflect that as well:

[This](https://github.com/home-assistant/core/pull/88256/files#diff-6ebd76c297e7ca0d88663399191136320ff278cf7ac4ea970d6745d26d96fa9cR314) line defines the test:
Given the time type is `CONF_ARRIVAL_TIME ` and is set to `test`
[this](https://github.com/home-assistant/core/pull/88256/files#diff-6ebd76c297e7ca0d88663399191136320ff278cf7ac4ea970d6745d26d96fa9cR339) line:
and the `user_input` does not contain `CONF_TIME` (because it was emptied)
[this](https://github.com/home-assistant/core/pull/88256/files#diff-6ebd76c297e7ca0d88663399191136320ff278cf7ac4ea970d6745d26d96fa9cR346) line:
then neither `CONF_ARRIVAL_TIME ` nor `CONF_DEPARTURE_TIME ` is configured

eifinger commented:
The other options can be emptied by selecting the value without any description. This looks horrible and needs to be addressed but not in this PR
<img width="576" alt="image" src="https://user-images.githubusercontent.com/1481961/229096999-72cad99f-2587-4d1d-92c4-81e92bb4806f.png">

epenet commented:
But what about `CONF_LANGUAGE`, `CONF_AVOID`, `CONF_TRAFFIC_MODEL`, `CONF_TRANSIT_MODE`, `CONF_TRANSIT_ROUTING_PREFERENCE`?

Since you dropped the default, I think they can't be reset.

eifinger commented:
They can still be reset. Just checked in manually again.
Why do you think we need the default for that?

epenet commented:
> You are right. Its no problem for entries which define less than 7 possible values. Because then radio buttons are used and I can select the one without any description aka `None` value. As soon as there are 7 or more possible values like in `CONF_LANGUAGE` `CONF_TRANSIT_MODE ` a combobox/dropdown(?) is used and reset does not work.

I suggest that you take a look at the discussion on #90412
An "easy" way to solve that is to set `default=""` or `default=None` in the schema, so that if frontend doesn't pass the key it gets set to the default (`None` or `""`).
That's what I started with there - but then it was improved with selectors, translations and a sentinel.

I suggest that you do something similar here.

epenet commented:
https://github.com/home-assistant/core/pull/90412 has been accepted and merged.
I suggest that you review it and adjust this one accordinly, with `selectors`, `translations` and a sentinel.
Maybe you should use selectors also for `waze`.

emontnemery commented on homeassistant/components/google_travel_time/config_flow.py:
```diff
@@ -40,6 +40,20 @@
 )
 from .helpers import InvalidApiKeyException, UnknownException, validate_config_entry
 
+OPTIONS_SCHEMA = vol.Schema(
+    {
+        vol.Optional(CONF_MODE): vol.In(TRAVEL_MODE),
+        vol.Optional(CONF_LANGUAGE): vol.In([None, *ALL_LANGUAGES]),
+        vol.Optional(CONF_AVOID): vol.In([None, *AVOID]),
+        vol.Optional(CONF_UNITS): vol.In(UNITS),
+        vol.Optional(CONF_TIME_TYPE): vol.In(TIME_TYPES),
+        vol.Optional(CONF_TIME): cv.string,
+        vol.Optional(CONF_TRAFFIC_MODEL): vol.In([None, *TRAVEL_MODEL]),
+        vol.Optional(CONF_TRANSIT_MODE): vol.In([None, *TRANSPORT_TYPE]),
+        vol.Optional(CONF_TRANSIT_ROUTING_PREFERENCE): vol.In([None, *TRANSIT_PREFS]),
+    }
+)
+
```
This should be modernized to use selectors instead.
I also agree with @epenet's comment

eifinger committed:
Use selectors

eifinger commented:
Added selectors, sentinel and translations. I did add a selector for the `time` field as this currently allows special configuration depending on `time_type`. This has to be split up in a separate PR

eifinger committed:
Use NONE_SENTINEL in tests

eifinger commented:
The OptionsFlow now looks like this:
<img width="584" alt="image" src="https://user-images.githubusercontent.com/1481961/230715830-c98103f6-4ca7-4b75-aa62-4e426cf69abc.png">

eifinger commented:
@emontnemery I addressed your review comments. Is there anything else to work on?
I would like to merge this so I can continue to work on some requested features on this integration.

edenhaus commented on homeassistant/components/google_travel_time/config_flow.py:
```diff
@@ -30,16 +35,82 @@
+        vol.Optional(CONF_TIME, default=""): cv.string,
+        vol.Optional(CONF_TRAFFIC_MODEL, default=NONE_SENTINEL): SelectSelector(
+            SelectSelectorConfig(
+                options=[NONE_SENTINEL] + sorted(TRAFFIC_MODEL),
```
```suggestion
                options=[NONE_SENTINEL] + sorted(TRAFFIC_MODELS),
```

I suggest renaming the constant name of the options to plural. Some are already in plural, others not.

eifinger committed:
apply suggestions

edenhaus commented on homeassistant/components/google_travel_time/const.py:
```diff
@@ -79,9 +79,9 @@
 
 AVOID = ["tolls", "highways", "ferries", "indoor"]
```
```suggestion
AVOIDS = ["tolls", "highways", "ferries", "indoor"]
```

eifinger committed:
rename AVOID to AVOID_OPTIONS

edenhaus commented:
Short update: Frontend will add support for not selecting any option, and therefore, we can remove all `NONE_SENTINEL`.
We should wait for the frontend before merging this PR

edenhaus commented on homeassistant/components/google_travel_time/config_flow.py:
```diff
@@ -30,16 +35,89 @@
 )
 from .helpers import InvalidApiKeyException, UnknownException, validate_config_entry
 
+NONE_SENTINEL = "none"
```
```suggestion
```

Similar to https://github.com/home-assistant/core/pull/101278
frontend PR will be merged after `2023.10.0` is released

edenhaus commented on homeassistant/components/google_travel_time/config_flow.py:
```diff
@@ -30,16 +35,89 @@
+                translation_key=CONF_LANGUAGE,
+            )
+        ),
+        vol.Optional(CONF_AVOID, default=NONE_SENTINEL): SelectSelector(
+            SelectSelectorConfig(
+                options=[NONE_SENTINEL] + AVOID_OPTIONS,
+                sort=True,
+                mode=SelectSelectorMode.DROPDOWN,
+                translation_key=CONF_AVOID,
+            )
```
```suggestion
        vol.Optional(CONF_AVOID): SelectSelector(
            SelectSelectorConfig(
                options=AVOID_OPTIONS,
                sort=True,
                mode=SelectSelectorMode.DROPDOWN,
                translation_key=CONF_AVOID,
            )
```

eifinger committed:
Remove NONE_SENTINEL

eifinger committed:
Update homeassistant/components/google_travel_time/strings.json

Co-authored-by: Robert Resch <robert@resch.dev>

eifinger committed:
Update homeassistant/components/google_travel_time/strings.json

Co-authored-by: Robert Resch <robert@resch.dev>

eifinger committed:
Update homeassistant/components/google_travel_time/strings.json

Co-authored-by: Robert Resch <robert@resch.dev>

eifinger committed:
Update homeassistant/components/google_travel_time/strings.json

Co-authored-by: Robert Resch <robert@resch.dev>

edenhaus reviewed (approved):
Thanks @eifinger :+1:
